crates/bevy_ecs/src/query/fetch.rs

```diff
@@ -12,94 +12,143 @@ pub use bevy_ecs_macros::WorldQuery;
+/// # Trait derivation
 ///
-/// The derive macro implements [`WorldQuery`] for your type and declares an additional struct
-/// which will be used as an item for query iterators. The implementation also generates two other
-/// structs that are used as [`WorldQuery::Fetch`](WorldQueryGats::Fetch) and
-/// [`WorldQuery::State`] associated types.
+/// Query design can be easily structured by deriving `WorldQuery` for custom types.
+/// Despite the added complexity, this approach has several advantages over using `WorldQuery` tuples.
+/// The most relevant improvements are:
 ///
-/// The derive macro requires every struct field to implement the `WorldQuery` trait.
+/// - Reusability across multiple systems.
+/// - There is no need to destructure a tuple since all fields are named.
+/// - Subqueries can be composed together to create a more complex query.
+/// - Methods can be implemented for the query items.
+/// - There is no hardcoded limit on the number of elements.
 ///
-/// **Note:** currently, the macro only supports named structs.
+/// This trait can only be derived if each field of the struct also implements `WorldQuery`.
+/// The derive macro only supports regular structs (structs with named fields).
 ///
 /// ```
 /// # use bevy_ecs::prelude::*;
 /// use bevy_ecs::query::WorldQuery;
-///
-/// #[derive(Component)]
-/// struct Foo;
-/// #[derive(Component)]
-/// struct Bar;
+/// #
+/// # #[derive(Component)]
+/// # struct ComponentA;
+/// # #[derive(Component)]
+/// # struct ComponentB;
 ///
 /// #[derive(WorldQuery)]
 /// struct MyQuery {
 ///     entity: Entity,
-///     // We must explicitly list out all lifetimes, as we are defining a struct
-///     foo: &'static Foo,
-///     bar: Option<&'static Bar>,
+///     // It is required that all reference lifetimes are explicitly annotated, just like in any
+///     // struct. Each lifetime should be 'static.
+///     component_a: &'static ComponentA,
+///     component_b: &'static ComponentB,
 /// }
 ///
 /// fn my_system(query: Query<MyQuery>) {
-///     for q in &query {
-///         // Note the type of the returned item.
-///         let q: MyQueryItem<'_> = q;
-///         q.foo;
+///     for q in query.iter() {
+///         q.component_a;
 ///     }
 /// }
-///
 /// # bevy_ecs::system::assert_is_system(my_system);
 /// ```
 ///
-/// ## Mutable queries
+/// ## Macro expansion
+///
+/// Expanding the macro will declare three or five additional structs, depending on whether or not the struct is marked as mutable.
+/// For a struct named `X`, the additional structs will be:
+///
+/// |Struct name|`mutable` only|Description|
+/// |:---:|:---:|---|
+/// |`XItem`|---|The type of the query item|
+/// |`XFetch`|---|Implements [`Fetch`] for `XItem`|
+/// |`XState`|---|Implements [`FetchState`]|
+/// |`XReadOnlyItem`|✓|The type of the read-only query item|
+/// |`XReadOnlyFetch`|✓|Implements [`Fetch`] for `XReadOnlyItem`|
+///
```

Comment:
```suggestion
/// ## Macro expansion
///
/// Expanding the macro will declare three or six additional structs, depending on whether or not the struct is marked as mutable.
/// For a struct named `X`, the additional structs will be:
///
/// |Struct name|`mutable` only|Description|
/// |:---:|:---:|---|
/// |`XState`|---|Used as the [`State`] type for `X` and `XReadOnly`|
/// |`XItem`|---|The type of the query item for `X`|
/// |`XFetch`|---|Used as the [`Fetch`] type for `X`|
/// |`XReadOnlyItem`|✓|The type of the query item for `XReadOnly`|
/// |`XReadOnlyFetch`|✓|Used as the [`Fetch`] type for `XReadOnly`|
/// |`XReadOnly`|✓|ReadOnly variant of `X`|
///
```
Although i am unsure if the `Fetch` and `State` types need to be explained so prominently along with the rest of the information (i.e. `XReadOnly` and `XReadOnlyItem`) since for the most users don't really have to care about these.